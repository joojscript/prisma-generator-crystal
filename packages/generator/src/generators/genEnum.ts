import { DMMF } from '@prisma/generator-helper'
import { StringHelpers } from '../helpers/strings'

export const genEnum = ({ name, values }: DMMF.DatamodelEnum) => {
  const enumValues = values
    .map(({ name }) => StringHelpers.capitalize(name))
    .join('\n  ')

  return `enum ${StringHelpers.capitalize(name)} 
  ${enumValues}
end`
}
