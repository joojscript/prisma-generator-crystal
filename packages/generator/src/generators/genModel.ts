import { DMMF } from '@prisma/generator-helper'
import { StringHelpers } from '../helpers/strings'

/**
 
 */

export const genModel = ({ name }: DMMF.Model) => {
  return 'Hello World'
}
