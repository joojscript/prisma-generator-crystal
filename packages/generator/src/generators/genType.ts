import { DMMF } from '@prisma/generator-helper'
import { StringHelpers } from '../helpers/strings'
import { TypeConvertions } from '../helpers/type-convertions'
import { Dictionary } from '@prisma/sdk'

export const genType = (
  { name, fields }: DMMF.Model,
  meta: Dictionary<string>,
) => {
  const typeName = StringHelpers.capitalize(name)

  const typeFields = fields
    .map(
      ({ name, type, isRequired, isList, hasDefaultValue, kind, ...rest }) => {
        const fieldName = StringHelpers.firstLowerCased(name)
        const fieldType = TypeConvertions.convert(type, kind) || ''
        const fieldRequired =
          fieldType && isRequired ? fieldType : `(${fieldType} | Nil)`
        const fieldList =
          fieldRequired && isList ? `Array(${fieldRequired})` : fieldRequired
        const fieldDefault =
          hasDefaultValue && TypeConvertions.fieldDefault(rest['default'])
            ? ` = ${TypeConvertions.fieldDefault(rest['default'])}`
            : ''

        return `property ${fieldName}: ${fieldList}${fieldDefault}`
      },
    )
    .join('\n  ')

  return `class ${typeName}Type
  ${typeFields}
end`
}
