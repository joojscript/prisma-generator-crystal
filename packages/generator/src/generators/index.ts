import { genModel } from '../generators/genModel'
import { genEnum } from '../generators/genEnum'
import { DMMF, GeneratorOptions } from '@prisma/generator-helper'

export const Generators: {
  [key: string]: (data: any, meta?: GeneratorOptions) => string
} = {
  models: genModel,
  enums: genEnum,
}
export default Generators
