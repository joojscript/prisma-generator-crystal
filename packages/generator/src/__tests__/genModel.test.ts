import { genModel } from '../../src/generators/genModel'
import { getSampleDMMF } from './__fixtures__/getSampleDMMF'

test('model generation', async () => {
  const sampleDMMF = await getSampleDMMF()

  sampleDMMF.datamodel.models.forEach((modelInfo) => {
    expect(genModel(modelInfo, {})).toMatchSnapshot(modelInfo.name)
  })
})
