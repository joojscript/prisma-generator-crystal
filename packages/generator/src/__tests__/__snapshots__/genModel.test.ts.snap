// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model generation: User 1`] = `
"class User
  property id: Int32
  property email: String
  property name: (String | Nil)
  property createdAt: Time
end"
`;
